IO - Estimated servo angles

MauroMombelli opened the pull request:
as in issue #925 we want in min and max the PWM value for a full 180°, and then minLimit and maxLimit are the physical limit (in degree) of movement.

Note: the limit use as 0° the center point of the servo

thenickdude commented:
Why is this desirable, exactly?

MauroMombelli commented:
build error seems in line with cleanflight/master, maybe you want this based on 1.8.1 so it (should) not fail?

hydra commented:
```
/src/main/config/config.c: In function 'writeEEPROM':

./src/main/config/config.c:855:5: error: size of unnamed array is negative

     BUILD_BUG_ON(sizeof(master_t) > FLASH_TO_RESERVE_FOR_CONFIG);

     ^
```
this means the configuration data required is greater than the flash space reserved for the configuration.

hydra commented:
Yesterday a change went in that changed the max servos from 8 to 10, thus the amount of storage required changed.  you additions take an extra 2 bytes per servo, 40 bytes total.

MauroMombelli commented:
@hydra i see, servo number should become target dependent, and this mean cleanflight should tell the configurator how many servo is gonna use.
is there something else? i may do it tomorrow if it ok

hydra commented:
@MauroMombelli  yes, i'm thinking that it should perhaps be target dependent too, since it's just a waste on targets that can't support that many servos.  this again comes down to #357 and #884.

The MSP protocol can already be used to figure out how many servos are supported by looking at the API version and the size of the packet returned by the servo related messages.  In fact i had to fix one MSP command yesterday when i changed the MAX_SEVOS to 10 from 8.

MauroMombelli commented:
@hydra using package size may result in bad thing, it is version dependent and may generate a lot of WTF.
Sending the plain number (uint8_t should be enough) followed by the structure maybe is a better idea

hydra commented:
ok, looking at this now

hydra commented on src/main/drivers/pwm_mapping.c:
```diff
@@ -77,35 +77,35 @@ enum {
+    PWM5  | (MAP_TO_MOTOR_OUTPUT << 8),  // Swap to servo if needed - naze TIM3
+    PWM6  | (MAP_TO_MOTOR_OUTPUT << 8),  // Swap to servo if needed - naze TIM3
+    PWM7  | (MAP_TO_MOTOR_OUTPUT << 8),  // Swap to servo if needed - naze TIM3
+    PWM8  | (MAP_TO_MOTOR_OUTPUT << 8),  // Swap to servo if needed - naze TIM3
```
although the information regarding timers is useful to developers it will go stale over time - comments cannot be refactored and should be avoided where possible.

MauroMombelli replied:
can we have a page on the wiki with this data for every board? it take a lot of time to check on the MCU datasheet

hydra replied:
look in timer.c. there are also plans to cleanup io/resource allocation.  no, no wiki page, we want it in clean code so it can be refactored.

hydra replied:
See https://github.com/cleanflight/cleanflight/issues/884

hydra commented on src/main/io/serial_msp.c:
```diff
@@ -1436,6 +1439,14 @@ static bool processInCommand(void)
 #endif
         break;
+    case MSP_SET_SERVO_LIMIT:
+#ifdef USE_SERVOS
```
at first glace i can't see why this is needed since MSP_SERVO_CONF has also been updated and seems correct.

hydra commented:
The MSP api version needs bumping when any MSP command is added/changed.

hydra commented on src/main/flight/mixer.h:
```diff
@@ -97,6 +97,8 @@ typedef struct servoParam_t {
     int16_t middle;                         // servo middle
     int8_t rate;                            // range [-100;+100] ; can be used to ajust a rate 0-100% and a direction
+    int8_t minLimit;                        // range [-90;90] ; can be used to adjust a rate 0-100% and a direction
+    int8_t maxLimit;                        // range [-90;90] ; can be used to adjust a rate 0-100% and a direction
```
with reference to configurator pull request https://github.com/cleanflight/cleanflight-configurator/pull/193

the comments don't match the usage and the variable names don't match the values!

```
                            'minLimit': data.getInt8(i + 7), //degree -90 90
                            'maxLimit': data.getInt8(i + 8) //degree -90 90
```
perhaps these should be called something like `minAngle` / `maxAngle`

hydra replied:
also, i think using uint8_t and making sure they are always positive numbers would be better, since the 'rate' can be used to reverse the direction.

hydra replied:
to clarify, say the min limit was -45 and the max limit was +45, what would happen if the user got the values for max and max transposed and specfied min +45 and max -45?

hydra replied:
after a bit more thought, i think the names should be `angleAtMin` and `angleAtMax`

hydra commented:
So, further to this. This PR is currently a bit weird since it has no purpose, all it does is accept input from the user and store it, the values are never actually used anywhere.

I know the goal is to use the values for the servo tilting mechanisms but if the code in configurator pull request https://github.com/cleanflight/cleanflight-configurator/pull/19 was merged then the user would likely be confused if they were setting up a pan/tilt or tricopter servo since the minLimit and maxLimit are never used.

I think the same settings should be applied to ALL servos, not just when they're used for the tilt mechanism.

hydra commented:
I cleaned up what was in this PR and created a new branch:

https://github.com/cleanflight/cleanflight/compare/servo-angles?expand=1

Please create yourself a new local branch from this branch and rebase this PR on this new branch with any additional changes.

MauroMombelli commented:
will fix those issue tomorrow (just tell me what to do about the travis notification)

>I think the same settings should be applied to ALL servos, not just when they're used for the tilt mechanism.

i think so, and also the gimbal code and probably the plane/wing code.

mine suggestion is to make the "writeServo" using us disappear as much as possible and use instead a more general version with use angles, even if will take a bit before updating all the code.

other trick would be to change the servo gui only for tilting arm and slowly change it also for what get ported to the new system.

>So, further to this. This PR is currently a bit weird since it has no purpose, all it does is accept input from the user and store it, the values are never actually used anywhere.

i wanted to break the pull request in smaller logical pieces as it may help taking care of them; the other pull request #872 is the complete one

hydra commented:
yes. ok. so what we could do is this:

1) merge the branch i created - even though it doesn't actually do anything yet, but serves as the basis for new code.
2) update the configurator PR so that angleAtMin and angleAtMax are left at defaults for all servos not related to TILT and the GUI should disable the fields in the servo list.  It should still show and backup/restore those values for ALL servos.
3) configurator's backup/restore configuration migration and handling of old API versions should be addressed.

what do you think?

MauroMombelli commented:
hi, i think this is a good idea, just the fact that the new branch is based on master give me problem as I have to backport on 1.8.1 or 1.9.0 for testing, as master cannot be trusted. (last time i tried wasn't working)

1. i merged the tilting arm pull request with your branch, updated the logic, and created the function radiansWriteServo similar to pwmWriteServo, witch can be used from others part of the code. I have not included the reverse logic into that, because there are multiple bit for every servo that indicate direction, and i don't understand why (i mean, a servo can move in only one direction right?). (see servoDirection() in flight/mxer.c)

2. 
>update the configurator PR

done but i don't have a naze so it live on th branch tilting on inkonova repo, after test will be merged into the pull-request branch

> It should still show and backup/restore those values for ALL servos.

I don't think i had changed any logic in servo tab, so i can't really understand hat you want here

3. working on that right now

hydra commented:
The previous MSP_SET_SERVO_CONF code would have been broken due to the value of INBUF_SIZE (even after my cleanups).  I made some changes to master today and merged them into the servo-angles branch.  I will close this PR.  The other PR's need to be rebased on the updated 'servo-angles' branch.

MauroMombelli commented:
@hydra  what is your fix? i have to increase the INBUF_SIZE to 72, but then when i arm the arming light start blinking and the esc reset.

I'm working on restoring the old two MPU comman system, but we have to keep an eyes on ram usage; the new 2 bite make 16 byte for the structure, plus 8 byte to the buffer increase, it means 24 byte to go over the ram limit for the naze, witch is a bit scary

maybe the servo/engine structure should be initialized at the startup based on configuration used, or different build can be provided for plane and quad.

really i don't know, let me know
